### PR TITLE
Update selection of type 3 response in ntlm.c

### LINKED ISSUE #3286: NTLMv2 not fully functional
### I did this

  I used the latest curl Version to authenticate against a porxy with --proxy-ntlm. The proxy was joined to AD which only accepts secure NTLMv2.
  
  Internet Explorer  NTLM exchange looks like:
IE NTLM type 1

Proxy-Authorization: NTLM TlRMTVNTUAABAAAAB4IIogAAAAAAAAAAAAAAAAAAAAAGAbEdAAAADw==\r\n
    NTLM Secure Service Provider
        NTLMSSP identifier: NTLMSSP
        NTLM Message Type: NTLMSSP_NEGOTIATE (0x00000001)
        Negotiate Flags: 0xa2088207, Negotiate 56, Negotiate 128, Negotiate Version, Negotiate Extended Security, Negotiate Always Sign, Negotiate NTLM key, Request Target, Negotiate OEM, Negotiate UNICODE
            1... .... .... .... .... .... .... .... = Negotiate 56: Set
            .0.. .... .... .... .... .... .... .... = Negotiate Key Exchange: Not set
            ..1. .... .... .... .... .... .... .... = Negotiate 128: Set
            ...0 .... .... .... .... .... .... .... = Negotiate 0x10000000: Not set
            .... 0... .... .... .... .... .... .... = Negotiate 0x08000000: Not set
            .... .0.. .... .... .... .... .... .... = Negotiate 0x04000000: Not set
            .... ..1. .... .... .... .... .... .... = Negotiate Version: Set
            .... ...0 .... .... .... .... .... .... = Negotiate 0x01000000: Not set
            .... .... 0... .... .... .... .... .... = Negotiate Target Info: Not set
            .... .... .0.. .... .... .... .... .... = Request Non-NT Session: Not set
            .... .... ..0. .... .... .... .... .... = Negotiate 0x00200000: Not set
            .... .... ...0 .... .... .... .... .... = Negotiate Identify: Not set
            .... .... .... 1... .... .... .... .... = Negotiate Extended Security: Set
            .... .... .... .0.. .... .... .... .... = Target Type Share: Not set
            .... .... .... ..0. .... .... .... .... = Target Type Server: Not set
            .... .... .... ...0 .... .... .... .... = Target Type Domain: Not set
            .... .... .... .... 1... .... .... .... = Negotiate Always Sign: Set
            .... .... .... .... .0.. .... .... .... = Negotiate 0x00004000: Not set
            .... .... .... .... ..0. .... .... .... = Negotiate OEM Workstation Supplied: Not set
            .... .... .... .... ...0 .... .... .... = Negotiate OEM Domain Supplied: Not set
            .... .... .... .... .... 0... .... .... = Negotiate Anonymous: Not set
            .... .... .... .... .... .0.. .... .... = Negotiate NT Only: Not set
            .... .... .... .... .... ..1. .... .... = Negotiate NTLM key: Set
            .... .... .... .... .... ...0 .... .... = Negotiate 0x00000100: Not set
            .... .... .... .... .... .... 0... .... = Negotiate Lan Manager Key: Not set
            .... .... .... .... .... .... .0.. .... = Negotiate Datagram: Not set
            .... .... .... .... .... .... ..0. .... = Negotiate Seal: Not set
            .... .... .... .... .... .... ...0 .... = Negotiate Sign: Not set
            .... .... .... .... .... .... .... 0... = Request 0x00000008: Not set
            .... .... .... .... .... .... .... .1.. = Request Target: Set
            .... .... .... .... .... .... .... ..1. = Negotiate OEM: Set
            .... .... .... .... .... .... .... ...1 = Negotiate UNICODE: Set
        Calling workstation domain: NULL
        Calling workstation name: NULL
        Version 6.1 (Build 7601); NTLM Current Revision 15
            Major Version: 6
            Minor Version: 1
            Build Number: 7601
            NTLM Current Revision: 15

IE(McAfee) NTLM type 2

Proxy-Authenticate: NTLM TlRMTVNTUAACAAAAAAAAAAAAAAAFgomgTynGuLy0QXwAAAAAAAAAAAAAAAAAAAAA\r\n
    NTLM Secure Service Provider
        NTLMSSP identifier: NTLMSSP
        NTLM Message Type: NTLMSSP_CHALLENGE (0x00000002)
        Target Name: NULL
        Negotiate Flags: 0xa0898205, Negotiate 56, Negotiate 128, Negotiate Target Info, Negotiate Extended Security, Target Type Domain, Negotiate Always Sign, Negotiate NTLM key, Request Target, Negotiate UNICODE
            1... .... .... .... .... .... .... .... = Negotiate 56: Set
            .0.. .... .... .... .... .... .... .... = Negotiate Key Exchange: Not set
            ..1. .... .... .... .... .... .... .... = Negotiate 128: Set
            ...0 .... .... .... .... .... .... .... = Negotiate 0x10000000: Not set
            .... 0... .... .... .... .... .... .... = Negotiate 0x08000000: Not set
            .... .0.. .... .... .... .... .... .... = Negotiate 0x04000000: Not set
            .... ..0. .... .... .... .... .... .... = Negotiate Version: Not set
            .... ...0 .... .... .... .... .... .... = Negotiate 0x01000000: Not set
            .... .... 1... .... .... .... .... .... = Negotiate Target Info: Set
            .... .... .0.. .... .... .... .... .... = Request Non-NT Session: Not set
            .... .... ..0. .... .... .... .... .... = Negotiate 0x00200000: Not set
            .... .... ...0 .... .... .... .... .... = Negotiate Identify: Not set
            .... .... .... 1... .... .... .... .... = Negotiate Extended Security: Set
            .... .... .... .0.. .... .... .... .... = Target Type Share: Not set
            .... .... .... ..0. .... .... .... .... = Target Type Server: Not set
            .... .... .... ...1 .... .... .... .... = Target Type Domain: Set
            .... .... .... .... 1... .... .... .... = Negotiate Always Sign: Set
            .... .... .... .... .0.. .... .... .... = Negotiate 0x00004000: Not set
            .... .... .... .... ..0. .... .... .... = Negotiate OEM Workstation Supplied: Not set
            .... .... .... .... ...0 .... .... .... = Negotiate OEM Domain Supplied: Not set
            .... .... .... .... .... 0... .... .... = Negotiate Anonymous: Not set
            .... .... .... .... .... .0.. .... .... = Negotiate NT Only: Not set
            .... .... .... .... .... ..1. .... .... = Negotiate NTLM key: Set
            .... .... .... .... .... ...0 .... .... = Negotiate 0x00000100: Not set
            .... .... .... .... .... .... 0... .... = Negotiate Lan Manager Key: Not set
            .... .... .... .... .... .... .0.. .... = Negotiate Datagram: Not set
            .... .... .... .... .... .... ..0. .... = Negotiate Seal: Not set
            .... .... .... .... .... .... ...0 .... = Negotiate Sign: Not set
            .... .... .... .... .... .... .... 0... = Request 0x00000008: Not set
            .... .... .... .... .... .... .... .1.. = Request Target: Set
            .... .... .... .... .... .... .... ..0. = Negotiate OEM: Not set
            .... .... .... .... .... .... .... ...1 = Negotiate UNICODE: Set
        NTLM Server Challenge: 4f29c6b8bcb4417c
        Reserved: 0000000000000000
        Target Info List: Empty

IE NTLM type 3

 [truncated]Proxy-Authorization: NTLM TlRMTVNTUAADAAAAGAAYAIQAAABoAGgAnAAAAAYABgBYAAAADAAMAF4AAAAaABo......
    NTLM Secure Service Provider
        NTLMSSP identifier: NTLMSSP
        NTLM Message Type: NTLMSSP_AUTH (0x00000003)
        Lan Manager Response: c5e5415ee4abe5e53111eb4c57ca317d6efc344bad2542a6
            Length: 24
            Maxlen: 24
            Offset: 132
        NTLM Response: 785714e1caa79f519213e4ae82d2f9b50101000000000000...
            Length: 104
            Maxlen: 104
            Offset: 156
            NTLMv2 Response: 785714e1caa79f519213e4ae82d2f9b50101000000000000...
                NTProofStr: 785714e1caa79f519213e4ae82d2f9b5
                Response Version: 1
                Hi Response Version: 1
                Z: 000000000000
                Time: Nov  8, 2018 16:58:10.360678400 UTC
                NTLMv2 Client Challenge: 6efc344bad2542a6
                Z: 00000000
                Attribute: Restrictions
                    NTLMV2 Response Item Type: Restrictions (0x0008)
                    NTLMV2 Response Item Length: 48
                    Restrictions: 30000000000000000000000000200000856e7b86caf2ee6f...
                Attribute: End of list
                    NTLMV2 Response Item Type: End of list (0x0000)
                    NTLMV2 Response Item Length: 0
                Z: 00000000
        Domain name: XXX
            Length: 6
            Maxlen: 6
            Offset: 88
        User name: user01
            Length: 12
            Maxlen: 12
            Offset: 94
        Host name: DESKTOP000001
            Length: 26
            Maxlen: 26
            Offset: 106
        Session Key: Empty
        Negotiate Flags: 0xa2888205, Negotiate 56, Negotiate 128, Negotiate Version, Negotiate Target Info, Negotiate Extended Security, Negotiate Always Sign, Negotiate NTLM key, Request Target, Negotiate UNICODE
            1... .... .... .... .... .... .... .... = Negotiate 56: Set
            .0.. .... .... .... .... .... .... .... = Negotiate Key Exchange: Not set
            ..1. .... .... .... .... .... .... .... = Negotiate 128: Set
            ...0 .... .... .... .... .... .... .... = Negotiate 0x10000000: Not set
            .... 0... .... .... .... .... .... .... = Negotiate 0x08000000: Not set
            .... .0.. .... .... .... .... .... .... = Negotiate 0x04000000: Not set
            .... ..1. .... .... .... .... .... .... = Negotiate Version: Set
            .... ...0 .... .... .... .... .... .... = Negotiate 0x01000000: Not set
            .... .... 1... .... .... .... .... .... = Negotiate Target Info: Set
            .... .... .0.. .... .... .... .... .... = Request Non-NT Session: Not set
            .... .... ..0. .... .... .... .... .... = Negotiate 0x00200000: Not set
            .... .... ...0 .... .... .... .... .... = Negotiate Identify: Not set
            .... .... .... 1... .... .... .... .... = Negotiate Extended Security: Set
            .... .... .... .0.. .... .... .... .... = Target Type Share: Not set
            .... .... .... ..0. .... .... .... .... = Target Type Server: Not set
            .... .... .... ...0 .... .... .... .... = Target Type Domain: Not set
            .... .... .... .... 1... .... .... .... = Negotiate Always Sign: Set
            .... .... .... .... .0.. .... .... .... = Negotiate 0x00004000: Not set
            .... .... .... .... ..0. .... .... .... = Negotiate OEM Workstation Supplied: Not set
            .... .... .... .... ...0 .... .... .... = Negotiate OEM Domain Supplied: Not set
            .... .... .... .... .... 0... .... .... = Negotiate Anonymous: Not set
            .... .... .... .... .... .0.. .... .... = Negotiate NT Only: Not set
            .... .... .... .... .... ..1. .... .... = Negotiate NTLM key: Set
            .... .... .... .... .... ...0 .... .... = Negotiate 0x00000100: Not set
            .... .... .... .... .... .... 0... .... = Negotiate Lan Manager Key: Not set
            .... .... .... .... .... .... .0.. .... = Negotiate Datagram: Not set
            .... .... .... .... .... .... ..0. .... = Negotiate Seal: Not set
            .... .... .... .... .... .... ...0 .... = Negotiate Sign: Not set
            .... .... .... .... .... .... .... 0... = Request 0x00000008: Not set
            .... .... .... .... .... .... .... .1.. = Request Target: Set
            .... .... .... .... .... .... .... ..0. = Negotiate OEM: Not set
            .... .... .... .... .... .... .... ...1 = Negotiate UNICODE: Set
        Version 6.1 (Build 7601); NTLM Current Revision 15
            Major Version: 6
            Minor Version: 1
            Build Number: 7601
            NTLM Current Revision: 15
        MIC: 87dec4750a7b6c9e0a8e9d8795a6cb94

Curl NTLM exchange is:

curl NTLM type 1

Proxy-Authorization: NTLM TlRMTVNTUAABAAAABoIIAAAAAAAAAAAAAAAAAAAAAAA=\r\n
    NTLM Secure Service Provider
        NTLMSSP identifier: NTLMSSP
        NTLM Message Type: NTLMSSP_NEGOTIATE (0x00000001)
        Negotiate Flags: 0x00088206, Negotiate Extended Security, Negotiate Always Sign, Negotiate NTLM key, Request Target, Negotiate OEM
            0... .... .... .... .... .... .... .... = Negotiate 56: Not set
            .0.. .... .... .... .... .... .... .... = Negotiate Key Exchange: Not set
            ..0. .... .... .... .... .... .... .... = Negotiate 128: Not set
            ...0 .... .... .... .... .... .... .... = Negotiate 0x10000000: Not set
            .... 0... .... .... .... .... .... .... = Negotiate 0x08000000: Not set
            .... .0.. .... .... .... .... .... .... = Negotiate 0x04000000: Not set
            .... ..0. .... .... .... .... .... .... = Negotiate Version: Not set
            .... ...0 .... .... .... .... .... .... = Negotiate 0x01000000: Not set
            .... .... 0... .... .... .... .... .... = Negotiate Target Info: Not set
            .... .... .0.. .... .... .... .... .... = Request Non-NT Session: Not set
            .... .... ..0. .... .... .... .... .... = Negotiate 0x00200000: Not set
            .... .... ...0 .... .... .... .... .... = Negotiate Identify: Not set
            .... .... .... 1... .... .... .... .... = Negotiate Extended Security: Set
            .... .... .... .0.. .... .... .... .... = Target Type Share: Not set
            .... .... .... ..0. .... .... .... .... = Target Type Server: Not set
            .... .... .... ...0 .... .... .... .... = Target Type Domain: Not set
            .... .... .... .... 1... .... .... .... = Negotiate Always Sign: Set
            .... .... .... .... .0.. .... .... .... = Negotiate 0x00004000: Not set
            .... .... .... .... ..0. .... .... .... = Negotiate OEM Workstation Supplied: Not set
            .... .... .... .... ...0 .... .... .... = Negotiate OEM Domain Supplied: Not set
            .... .... .... .... .... 0... .... .... = Negotiate Anonymous: Not set
            .... .... .... .... .... .0.. .... .... = Negotiate NT Only: Not set
            .... .... .... .... .... ..1. .... .... = Negotiate NTLM key: Set
            .... .... .... .... .... ...0 .... .... = Negotiate 0x00000100: Not set
            .... .... .... .... .... .... 0... .... = Negotiate Lan Manager Key: Not set
            .... .... .... .... .... .... .0.. .... = Negotiate Datagram: Not set
            .... .... .... .... .... .... ..0. .... = Negotiate Seal: Not set
            .... .... .... .... .... .... ...0 .... = Negotiate Sign: Not set
            .... .... .... .... .... .... .... 0... = Request 0x00000008: Not set
            .... .... .... .... .... .... .... .1.. = Request Target: Set
            .... .... .... .... .... .... .... ..1. = Negotiate OEM: Set
            .... .... .... .... .... .... .... ...0 = Negotiate UNICODE: Not set
        Calling workstation domain: NULL
        Calling workstation name: NULL

curl (McAfee) NTLM type 2

Proxy-Authenticate: NTLM TlRMTVNTUAACAAAAAAAAAAAAAAAGgokAXmf7HuqyuFEAAAAAAAAAAAAAAAAAAAAA\r\n
    NTLM Secure Service Provider
        NTLMSSP identifier: NTLMSSP
        NTLM Message Type: NTLMSSP_CHALLENGE (0x00000002)
        Target Name: NULL
        Negotiate Flags: 0x00898206, Negotiate Target Info, Negotiate Extended Security, Target Type Domain, Negotiate Always Sign, Negotiate NTLM key, Request Target, Negotiate OEM
            0... .... .... .... .... .... .... .... = Negotiate 56: Not set
            .0.. .... .... .... .... .... .... .... = Negotiate Key Exchange: Not set
            ..0. .... .... .... .... .... .... .... = Negotiate 128: Not set
            ...0 .... .... .... .... .... .... .... = Negotiate 0x10000000: Not set
            .... 0... .... .... .... .... .... .... = Negotiate 0x08000000: Not set
            .... .0.. .... .... .... .... .... .... = Negotiate 0x04000000: Not set
            .... ..0. .... .... .... .... .... .... = Negotiate Version: Not set
            .... ...0 .... .... .... .... .... .... = Negotiate 0x01000000: Not set
            .... .... 1... .... .... .... .... .... = Negotiate Target Info: Set
            .... .... .0.. .... .... .... .... .... = Request Non-NT Session: Not set
            .... .... ..0. .... .... .... .... .... = Negotiate 0x00200000: Not set
            .... .... ...0 .... .... .... .... .... = Negotiate Identify: Not set
            .... .... .... 1... .... .... .... .... = Negotiate Extended Security: Set
            .... .... .... .0.. .... .... .... .... = Target Type Share: Not set
            .... .... .... ..0. .... .... .... .... = Target Type Server: Not set
            .... .... .... ...1 .... .... .... .... = Target Type Domain: Set
            .... .... .... .... 1... .... .... .... = Negotiate Always Sign: Set
            .... .... .... .... .0.. .... .... .... = Negotiate 0x00004000: Not set
            .... .... .... .... ..0. .... .... .... = Negotiate OEM Workstation Supplied: Not set
            .... .... .... .... ...0 .... .... .... = Negotiate OEM Domain Supplied: Not set
            .... .... .... .... .... 0... .... .... = Negotiate Anonymous: Not set
            .... .... .... .... .... .0.. .... .... = Negotiate NT Only: Not set
            .... .... .... .... .... ..1. .... .... = Negotiate NTLM key: Set
            .... .... .... .... .... ...0 .... .... = Negotiate 0x00000100: Not set
            .... .... .... .... .... .... 0... .... = Negotiate Lan Manager Key: Not set
            .... .... .... .... .... .... .0.. .... = Negotiate Datagram: Not set
            .... .... .... .... .... .... ..0. .... = Negotiate Seal: Not set
            .... .... .... .... .... .... ...0 .... = Negotiate Sign: Not set
            .... .... .... .... .... .... .... 0... = Request 0x00000008: Not set
            .... .... .... .... .... .... .... .1.. = Request Target: Set
            .... .... .... .... .... .... .... ..1. = Negotiate OEM: Set
            .... .... .... .... .... .... .... ...0 = Negotiate UNICODE: Not set
        NTLM Server Challenge: 5e67fb1eeab2b851
        Reserved: 0000000000000000
        Target Info List: Empty

Curl type 3

Proxy-Authorization: NTLM TlRMTVNTUAADAAAAGAAYAEAAAAAYABgAWAAAAAMAAwBwAAAABg......
    NTLM Secure Service Provider
        NTLMSSP identifier: NTLMSSP
        NTLM Message Type: NTLMSSP_AUTH (0x00000003)
        Lan Manager Response: a70028552819d71e00000000000000000000000000000000
            Length: 24
            Maxlen: 24
            Offset: 64
        LMv2 Client Challenge: a70028552819d71e
        NTLM Response: a7b5c28a0e8d60b9dc12ba0cef407f562347f71d4fd15234
            Length: 24
            Maxlen: 24
            Offset: 88
        Domain name: XXX
            Length: 3
            Maxlen: 3
            Offset: 112
        User name: user01
            Length: 6
            Maxlen: 6
            Offset: 115
        Host name: host01
            Length: 6
            Maxlen: 6
            Offset: 121
        Session Key: Empty
        Negotiate Flags: 0x00898206, Negotiate Target Info, Negotiate Extended Security, Target Type Domain, Negotiate Always Sign, Negotiate NTLM key, Request Target, Negotiate OEM
            0... .... .... .... .... .... .... .... = Negotiate 56: Not set
            .0.. .... .... .... .... .... .... .... = Negotiate Key Exchange: Not set
            ..0. .... .... .... .... .... .... .... = Negotiate 128: Not set
            ...0 .... .... .... .... .... .... .... = Negotiate 0x10000000: Not set
            .... 0... .... .... .... .... .... .... = Negotiate 0x08000000: Not set
            .... .0.. .... .... .... .... .... .... = Negotiate 0x04000000: Not set
            .... ..0. .... .... .... .... .... .... = Negotiate Version: Not set
            .... ...0 .... .... .... .... .... .... = Negotiate 0x01000000: Not set
            .... .... 1... .... .... .... .... .... = Negotiate Target Info: Set
            .... .... .0.. .... .... .... .... .... = Request Non-NT Session: Not set
            .... .... ..0. .... .... .... .... .... = Negotiate 0x00200000: Not set
            .... .... ...0 .... .... .... .... .... = Negotiate Identify: Not set
            .... .... .... 1... .... .... .... .... = Negotiate Extended Security: Set
            .... .... .... .0.. .... .... .... .... = Target Type Share: Not set
            .... .... .... ..0. .... .... .... .... = Target Type Server: Not set
            .... .... .... ...1 .... .... .... .... = Target Type Domain: Set
            .... .... .... .... 1... .... .... .... = Negotiate Always Sign: Set
            .... .... .... .... .0.. .... .... .... = Negotiate 0x00004000: Not set
            .... .... .... .... ..0. .... .... .... = Negotiate OEM Workstation Supplied: Not set
            .... .... .... .... ...0 .... .... .... = Negotiate OEM Domain Supplied: Not set
            .... .... .... .... .... 0... .... .... = Negotiate Anonymous: Not set
            .... .... .... .... .... .0.. .... .... = Negotiate NT Only: Not set
            .... .... .... .... .... ..1. .... .... = Negotiate NTLM key: Set
            .... .... .... .... .... ...0 .... .... = Negotiate 0x00000100: Not set
            .... .... .... .... .... .... 0... .... = Negotiate Lan Manager Key: Not set
            .... .... .... .... .... .... .0.. .... = Negotiate Datagram: Not set
            .... .... .... .... .... .... ..0. .... = Negotiate Seal: Not set
            .... .... .... .... .... .... ...0 .... = Negotiate Sign: Not set
            .... .... .... .... .... .... .... 0... = Request 0x00000008: Not set
            .... .... .... .... .... .... .... .1.. = Request Target: Set
            .... .... .... .... .... .... .... ..1. = Negotiate OEM: Set
            .... .... .... .... .... .... .... ...0 = Negotiate UNICODE: Not set

 

The difference can be seen in the typ3 response

IE:

        NTLM Response: 785714e1caa79f519213e4ae82d2f9b50101000000000000...
            Length: 104
            Maxlen: 104
            Offset: 156
            NTLMv2 Response: 785714e1caa79f519213e4ae82d2f9b50101000000000000...
Curl:
        Lan Manager Response: a70028552819d71e00000000000000000000000000000000
            Length: 24
            Maxlen: 24
            Offset: 64
        LMv2 Client Challenge: a70028552819d71e
       
A change in lib/vauth/ntlm.c  

 
#if defined(USE_NTRESPONSES) && defined(USE_NTLM_V2)
//MM No idea why this works, but it does
//   if(ntlm->target_info_len) {
  if(ntlm->flags & NTLMFLAG_NEGOTIATE_TARGET_INFO ) {
    unsigned char ntbuffer[0x18];
    unsigned char entropy[8];

### I expected the following

I expected successful authentication, but received a rejection.

### curl/libcurl version

> src/curl -V
curl 7.63.0-DEV (i686-suse-linux-gnu) libcurl/7.63.0-DEV OpenSSL/1.0.1g zlib/1.2.7 libssh2/1.4.3
Release-Date: [unreleased]
Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps pop3 pop3s rtsp scp sftp smb smbs smtp smtps telnet tftp 
Features: AsynchDNS Largefile GSS-API Kerberos SPNEGO NTLM NTLM_WB SSL libz TLS-SRP UnixSockets HTTPS-proxy 

### operating system

SLES 11

### PR DESCRIPTION
**Pull request by huaraz**
NTLM2 did not work i.e. no NTLMv2 response was created. Changing the check seems to work. 

https://winprotocoldoc.blob.core.windows.net/productionwindowsarchives/MS-NLMP/[MS-NLMP].pdf

Fixes #3286

### CODE REVIEW

**Comment by huaraz** [comment]
Does the Travis-CI log indicate NTLM failures ?  It is hard to read/understand the result in the log. 

Markus

**Comment by bagder** [comment]
These test cases failed: 2025 2028 2029 2030 2031

Yes, those tests seem to fail now and they all involve NTLM...

**Comment by huaraz** [comment]
I feared that be the case.  Would it be possible to get the capture of the NTLM exchange i.e. what were the flags in type 1,2 and 3 for the successful and failed auth attempts ? 

Is there an option to test against an AD server which has NTLMv1 / LM disabled ?

**Comment by bagder** [comment]
> Is there an option to test against an AD server which has NTLMv1 / LM disabled ?

All curl tests are run either without a test server or with a "synthetic" test server that we build ourselves. This way everyone can build and run the tests and they can run in the CI etc without having to connect anywhere or need any magic accounts. (Also, most curl developers are on non-windows platforms.)

For this particular failure, you've updated bits that curl sends in its NTLM headers without updating the corresponding tests so this is not a surprise. The tests will verify that curl sends exactly what is expected of it!

**Comment by huaraz** [comment]
The change shouldn't break anything for a "old" "insecure" AD server, but add additional functionality to support AD server which have LM and NTLMv1 disabled. 

So if the tests fail I need to understand what the synthetic setup looks like to adjust as I have a limited test environment.  Is that possible ? 

Thank you
Markus

**Comment by bagder** [comment]
Starting with the first test fail, [test2025](https://travis-ci.org/curl/curl/jobs/456614510#L5941), I think the critical diff is this:

~~~diff
-Authorization: NTLM TlRMTVNTUAADAAAAGAAYAEAAAAAYABgAWAAAAAAAAABwAAAACAAIAHAAAAAIAAgAeAAAAAAAAAAAAAAABoIBAI+/Fp9IERAQ74OsdNPbBpg7o8CVwLSO4DtFyIcZHUMKVktWIu92s2892OVpd2JzqnRlc3R1c2VyY3VybGhvc3Q=[CR][LF]
+Authorization: NTLM TlRMTVNTUAADAAAAGAAYAEAAAAAYABgAWAAAAAAAAABwAAAACAAIAHAAAAAIAAgAeAAAAAAAAAAAAAAABoIBACSb60vUMLShAAAAAAAAAAAAAAAAAAAAAHSb9yQXjz3gIcUWePywK8aoyXJ4KokJ13Rlc3R1c2VyY3VybGhvc3Q=[CR][LF]
~~~
If you paste them into a text editor you can see how the headers start differ at position 105. Which isn't a surprise since the code has changed how the `Authorization:` header is created without updating the tests...

**Comment by bagder** [comment]
They do both, but yes I'm saying these five tests all report the string differences as failures. Since they get a different string than they expect!

**Comment by huaraz** [comment]
So that means basically the change works. The tests need to be only adjusted ?

**Comment by bagder** [comment]
If you say so. The custom test server we run really doesn't "know" NTLM, it just plays back whatever we tell it to.

If you want to verify that your code really works and speaks NTLMv2 I trust that you use it against a real server that implements and supports that protocol. Once you know your code works against such a server, then simply updating the strings in the corresponding curl tests should be fine and will make sure we don't regress this going forward.

File: lib/vauth/ntlm.c [hunk 1]
```diff
@@ -451,6 +451,7 @@ CURLcode Curl_auth_create_ntlm_type1_message(struct Curl_easy *data,
             LONGQUARTET(NTLMFLAG_NEGOTIATE_OEM |
                         NTLMFLAG_REQUEST_TARGET |
                         NTLMFLAG_NEGOTIATE_NTLM_KEY |
+                        NTLMFLAG_NEGOTIATE_NTLM2_KEY |
                         NTLM2FLAG |
```
Comment (mkauf) [inline lib/vauth/ntlm.c]: ``NTLM2FLAG`` already contains ``NTLMFLAG_NEGOTIATE_NTLM2_KEY``, see line ~408:
```
#if defined(USE_NTRESPONSES) && defined(USE_NTLM2SESSION)
#define NTLM2FLAG NTLMFLAG_NEGOTIATE_NTLM2_KEY
#else
#define NTLM2FLAG 0
#endif
```

File: lib/vauth/ntlm.c [hunk 2]
```diff
@@ -599,7 +600,7 @@ CURLcode Curl_auth_create_ntlm_type3_message(struct Curl_easy *data,
 
 #if defined(USE_NTRESPONSES) && defined(USE_NTLM2SESSION)
   /* We don't support NTLM2 if we don't have USE_NTRESPONSES */
-  if(ntlm->flags & NTLMFLAG_NEGOTIATE_NTLM2_KEY) {
+  if(ntlm->flags & NTLMFLAG_NEGOTIATE_NTLM_KEY) {
```
Comment (mkauf) [inline lib/vauth/ntlm.c]: I don't understand this change. Does this mean that an NTLMv1 message is sent to servers that support NTLMv2 ?

File: lib/vauth/ntlm.c (hunk 1 above)
Comment (huaraz) [reply lib/vauth/ntlm.c]: Oh I overlooked that. Thank you

File: lib/vauth/ntlm.c (hunk 2 above)
Comment (huaraz) [reply lib/vauth/ntlm.c]: That is the part of the code I am not sure about.  The comment is about NTLM2, but the previous section is really the NTLMv2 section i.e. where the correct response is created.  This one seem to create only the LM response.  ( At least the was the result from the wireshark captures) 

Markus

**Comment by huaraz** [comment]
Does anybody know which Negotiate flag combination reflects the 5 cases ?

There are five authentication levels.

Send LM & NTLM responses: Clients use LM and NTLM authentication, and never use NTLMv2 session security; DCs accept LM, NTLM, and NTLMv2 authentication.

Send LM & NTLM - use NTLMv2 session security if negotiated: Clients use LM and NTLM authentication, and use NTLMv2 session security if server supports it; DCs accept LM, NTLM, and NTLMv2 authentication.

Send NTLM response only: Clients use NTLM authentication only, and use NTLMv2 session security if server supports it; DCs accept LM, NTLM, and NTLMv2 authentication.

Send NTLMv2 response only: Clients use NTLMv2 authentication only, and use NTLMv2 session security if server supports it; DCs accept LM, NTLM, and NTLMv2 authentication.

Send NTLMv2 response only\refuse LM: Clients use NTLMv2 authentication only, and use NTLMv2 session security if server supports it; DCs refuse LM (accept only NTLM and NTLMv2 authentication).

Send NTLMv2 response only\refuse LM & NTLM: Clients use NTLMv2 authentication only, and use NTLMv2 session security if server supports it; DCs refuse LM and NTLM (accept only NTLMv2 authentication).

Markus

**Comment by huaraz** [comment]
How can I rerun the tests ? They don't seem to be related to my code changes 

Thank you
Markus

**Comment by MarcelRaad** [comment]
I've restarted the failed Travis tests.

**Comment by huaraz** [comment]
It looks like the master branch is broken 
 

make[2]: Entering directory `/home/travis/build/curl/curl/tests/unit'
  CC       unit1300-unit1300.o
  CC       ../libtest/unit1300-first.o
make[2]: *** No rule to make target `../../src/libcurltool.la', needed by `unit1300'.  Stop.
make[2]: Leaving directory `/home/travis/build/curl/curl/tests/unit'
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory `/home/travis/build/curl/curl/tests'
make: *** [test-nonflaky] Error 2

**Comment by MarcelRaad** [comment]
Actually, at least the first failure is because of:

```
./vauth/ntlm.c:634:73: warning: Trailing whitespace (TRAILINGSPACE)
     /* NTLM v2 session security is a misnomer because it is not NTLM v2. 
                                                                         ^
./vauth/ntlm.c:635:84: warning: Longer than 79 columns (LONGLINE)
        It is NTLM v1 using the extended session security that is also in NTLM v2 */
```

Building with `--enable-debug` (or running `checksrc` explicitly) will show you these.

**Comment by huaraz** [comment]
> Actually, at least the first failure is because of:
> 
> ```
> ./vauth/ntlm.c:634:73: warning: Trailing whitespace (TRAILINGSPACE)
>      /* NTLM v2 session security is a misnomer because it is not NTLM v2. 
>                                                                          ^
> ./vauth/ntlm.c:635:84: warning: Longer than 79 columns (LONGLINE)
>         It is NTLM v1 using the extended session security that is also in NTLM v2 */
> ```
> Building with `--enable-debug` (or running `checksrc` explicitly) will show you these.

Thank you for the hint.   I did not find this in the log.

**Comment by huaraz** [comment]
Not sure what the cause is.  Looks more like an issue with the testing environment. 

Markus

**Comment by bagder** [comment]
I restarted the job. Only administrators can restart the travis jobs unfortunately.

**Comment by huaraz** [comment]
Thank you.   So it looks like all tests run now successfully.  It should be now ready for the next step i.e. real world tests.

Markus

**Comment by MarcelRaad** [comment]
Cool! I'll test it as soon as I'm back at work next Friday.

**Comment by MarcelRaad** [comment]
This works fine for me.

@huaraz Could you please squash the commits and remove the commented out debug code?

**Comment by huaraz** [comment]
Sorry I am too new to git and it seems I misunderstood how "squashing"  works.  I deleted the patch and recreated a new pull

Markus

**Pull request closed without merging**